### ISSUE TITLE
Templates for standard radioList and checkboxList inputs

### ISSUE THREAD

**Issue by justinvoelker** [yiisoft/yii2#7448]
It would be great to have to ability to specify the template of standard radioList and checkboxList fields.  I have the desire/need to move the input outside of the label and while this is possible with template for checkboxes and radio buttons, it is not possible with radioList and checkboxList.  The BaseHtml helper functions for radio and checkbox (the functions responsible for each of the radio buttons or checkboxes in a list) are hardcoded as follows
```php
$content = static::label(static::input('radio', $name, $value, $options) . ' ' . $label, null, $labelOptions);
```
Ideally we would be able to specify if we want the input inside the label or before it.  Along with this would be the inclusion of the "for" attribute for the label and "id" for the input so they two can be linked together without being nested.  In an attempt to get the functionality I was after, I hacked the together the following replacement for the line above:
```php
$options['id'] = str_replace('[', '-', substr($name, 0, strlen($name)-1)) . '-' . $value;
$content = static::input('radio', $name, $value, $options) . static::label($label, $options['id'], $labelOptions);
```
Checkboxes are nearly identical except the substring subtracts 3 from the length and obviously the input type is 'checkbox.'

I'm sure about the best way to implement this while making it backward compatible and still allowing for the customization so I'm not submitting any pull requests.  If there are any ideas for how to achieve this with the base code as it stands today that would be great.  Otherwise, I would like to request an enhancement to make this possible.

**EDIT Feb 26, 2015**
The same is true for inline radioList and checkboxList.  The `{input}` is hardcoded to have the actual input inside a label with no way to separate the two.  Both radioList/checkboxList and their inline variations are using BaseHtml functions radio() and checkbox() to render the individual items which do not allow separation of the input from inside the label.

It seems like we might need to add configuration options for `checkbboxListTemplate` and `radioListTemplate` to cover the non-inline versions of these as well.

**Comment by mikehaertl** [comment]
In case you use `yii\bootstrap\ActiveField` you can already do that to some extend.

http://www.yiiframework.com/doc-2.0/yii-bootstrap-activefield.html

**Comment by justinvoelker** [comment]
I am indeed using `yii\bootstrap\ActiveField` but there is no ability to change the template for the individual inputs of a radioList or cbeckboxList.  If you look at the `$content` specified above you can see that there is no template option like radio or checkbox fields.  Instead, `static::input` is hard-coded to always be inside `static::label.`

**Comment by justinvoelker** [comment]
Edited the original issue since the same inability to separate the input from inside the label affects the inline variations of the lists as well.

**Comment by mikehaertl** [comment]
Have you seen that you can provide a callback in the [item option](http://www.yiiframework.com/doc-2.0/yii-bootstrap-activefield.html#radioList()-detail)? This way you should be able to render the list items in whatever way you want.

**Comment by justinvoelker** [comment]
Still doing plenty of learning I haven't worked with callbacks yet but a quick attempt at this one shows it is easy enough (and short enough) to add a quick function.  But, since radioList is calling Html::radio to actually output each item, I'd have to include/modify all of the radio() logic as well to swap that one piece.  Since I'm looking for an easily repeatable solution I don't want to have a couple dozen lines of code repeated for each list field.  Rather, if it sounds like this won't be addressed in the baseline framework I'll work on an extension that overrides radioList and checkboxList (along with the radio/checkbox Html helpers) to output them as needed.  Then a simple call to my class in ActiveForm and I'll have an easily repeatable solution.

**Comment by justinvoelker** [comment]
Here's my attempt at writing an extension to address the issues I was having. Rather than opening up new functionality I just addressed my problems and implemented the functionality I needed.  If the baseline functionality ever changes I will update my extension to use what is delivered. Until then, I made what I need.

**Comment by niklaswolf** [comment]
@justinvoelker I'm having exactly the same problem: I want to render the radio-inputs outside of the label. You mentioned an extension that you wrote, is it available to public?

**Comment by justinvoelker** [comment]
@niklaswolf Seems when I closed this I forgot to actually link to what I created. It is actually an extension to implement [Awesome Bootstrap Checkbox](https://github.com/flatlogic/awesome-bootstrap-checkbox/). So, not so much an extension to fix the problem but rather to address my exact need. Here is the extension I wrote in case it can help you out at all: [yii2-awesomebootstrapcheckbox](https://github.com/justinvoelker/yii2-awesomebootstrapcheckbox)
